shared/src/store/action-reducers/exercise.ts

```diff
@@ -171,3 +194,76 @@ function refreshTransfer(
         letElementArrive(draftState, type, element.id);
     });
 }
+
+export interface TreatmentAssignment {
+    [patientId: UUID]: { [personnelType in PersonnelType]: number };
+}
+
+function calculateTreatmentAssignment(
+    draftState: Mutable<ExerciseState>
+): TreatmentAssignment {
+    const treatmentAssignment = StrictObject.fromEntries(
+        Object.keys(draftState.patients).map((patientId) => [
+            patientId,
+            StrictObject.fromEntries(
+                StrictObject.keys(personnelTypeAllowedValues).map(
+                    (personnelType) => [personnelType, 0]
+                )
+            ),
+        ])
+    ) as TreatmentAssignment;
+
+    StrictObject.values(draftState.personnel).forEach((personnel) => {
+        const assignedPatientCount = StrictObject.keys(
+            personnel.assignedPatientIds
+        ).length;
+        StrictObject.keys(personnel.assignedPatientIds)
+            .filter((patientId) => treatmentAssignment[patientId])
+            .forEach((patientId) => {
+                treatmentAssignment[patientId]![personnel.personnelType]! +=
+                    1 / assignedPatientCount;
+            });
+    });
+
+    return treatmentAssignment;
+}
+
+function evaluateTreatmentReassignment(
+    draftState: Mutable<ExerciseState>,
+    newTreatmentAssignment: TreatmentAssignment
+) {
+    if (!draftState.previousTreatmentAssignment) return;
+
+    Object.keys(newTreatmentAssignment)
+        .filter((patientId) =>
+            StrictObject.keys(personnelTypeAllowedValues).some(
+                (personnelType) =>
+                    newTreatmentAssignment[patientId]![personnelType] !==
+                    draftState.previousTreatmentAssignment![patientId]?.[
+                        personnelType
+                    ]
+            )
+        )
+        .forEach((patientId) => {
+            logPatient(
+                draftState,
+                StrictObject.entries(newTreatmentAssignment[patientId]!)
+                    .filter(([, count]) => count > 0)
+                    .map(([personnelType]) =>
+                        createPersonnelTypeTag(draftState, personnelType)
+                    ),
+                `Diese Einsatzkräfte wurden dem Patienten neu zugeteilt: ${
+                    StrictObject.entries(newTreatmentAssignment[patientId]!)!
+                        .filter(([, count]) => count > 0)
+                        .map(
+                            ([personnelType, count]) =>
+                                `${+count.toFixed(2)} ${
+                                    personnelTypeNames[personnelType]
+                                }`
+                        )
+                        .join(', ') ?? 'Keine Einsatzkräfte'
```

Comment:
This doesn't work since `[].join()` does not return undefined.

Reply:
You're right